File: financial-clearing-house/src/python/netting_engine.py
"""
Netting Engine for the Clearing House.

Implements multilateral netting to minimize the number of actual money movements.
Uses a greedy creditor-debtor matching algorithm that achieves O(N-1) transfers
for N banks with non-zero positions.
"""
from collections import defaultdict
from decimal import Decimal
from typing import Dict, List, Tuple
from dataclasses import dataclass
import heapq

from models import Transaction, SettlementInstruction, SettlementBatch


@dataclass
class NetPosition:
    """
    Net position of a bank after processing all transactions.

    Positive = bank is a creditor (will receive money)
    Negative = bank is a debtor (must pay money)
    """
    bank: str
    amount: Decimal

    def is_creditor(self) -> bool:
        return self.amount > 0

    def is_debtor(self) -> bool:
        return self.amount < 0

    def is_flat(self) -> bool:
        return self.amount == 0


class NettingEngine:
    """
    Multilateral netting engine that minimizes settlement transfers.

    Algorithm:
    1. Calculate net position per bank (sum of all inflows - outflows)
    2. Partition banks into creditors (net > 0) and debtors (net < 0)
    3. Greedily match largest creditor with largest debtor
    4. Generate settlement instruction for the matched amount
    5. Update positions and repeat until all positions are zero

    This achieves the minimum number of transfers (N-1 for N non-zero banks)
    when there are no transfer fees to optimize for.
    """

    def calculate_net_positions(
        self, transactions: List[Transaction]
    ) -> Dict[str, Decimal]:
        """
        Calculate the net position for each bank.

        Args:
            transactions: List of transactions to process

        Returns:
            Dictionary mapping bank_id to net position
            (positive = receive, negative = pay)
        """
        positions: Dict[str, Decimal] = defaultdict(Decimal)

        for txn in transactions:
            # Payer loses money
            positions[txn.payer] -= txn.amount
            # Payee gains money
            positions[txn.payee] += txn.amount

        return dict(positions)

    def validate_positions(self, positions: Dict[str, Decimal]) -> bool:
        """
        Validate that all positions sum to zero (system is balanced).

        Raises:
            ValueError: If the system is unbalanced
        """
        total = sum(positions.values())
        if total != 0:
            raise ValueError(
                f"System unbalanced! Positions sum to {total}, expected 0. "
                "This indicates lost or created money."
            )
        return True

    def generate_settlements(
        self, positions: Dict[str, Decimal]
    ) -> List[SettlementInstruction]:
        """
        Generate minimal settlement instructions using greedy matching.

        Uses a heap-based approach to always match the largest creditor
        with the largest debtor for optimal efficiency.

        Args:
            positions: Net position per bank

        Returns:
            List of SettlementInstruction objects (minimal set)
        """
        self.validate_positions(positions)

        # Separate into creditors and debtors using max-heaps
        # Python heapq is min-heap, so negate values for max-heap behavior
        creditors: List[Tuple[Decimal, str]] = []
        debtors: List[Tuple[Decimal, str]] = []

        for bank, amount in positions.items():
            if amount > 0:
                # Negate for max-heap (largest creditor first)
                heapq.heappush(creditors, (-amount, bank))
            elif amount < 0:
                # Already negative, push as-is for max-heap (largest debtor first)
                heapq.heappush(debtors, (amount, bank))

        instructions: List[SettlementInstruction] = []

        while creditors and debtors:
            # Get largest creditor and debtor
            neg_credit, creditor = heapq.heappop(creditors)
            debt, debtor = heapq.heappop(debtors)

            credit_amount = -neg_credit  # Restore positive value
            debt_amount = -debt  # Convert to positive for comparison

            # Settlement amount is the minimum of the two
            settlement_amount = min(credit_amount, debt_amount)

            # Generate settlement instruction
            instruction = SettlementInstruction(
                payer=debtor,
                payee=creditor,
                amount=settlement_amount
            )
            instructions.append(instruction)

            # Update remaining positions
            remaining_credit = credit_amount - settlement_amount
            remaining_debt = debt_amount - settlement_amount

            if remaining_credit > 0:
                heapq.heappush(creditors, (-remaining_credit, creditor))
            if remaining_debt > 0:
                heapq.heappush(debtors, (-remaining_debt, debtor))

        return instructions

    def run_settlement(
        self, transactions: List[Transaction], batch_id: str = None
    ) -> SettlementBatch:
        """
        Run the full settlement process for a list of transactions.

        Args:
            transactions: List of transactions to settle
            batch_id: Optional batch identifier

        Returns:
            SettlementBatch containing positions and instructions
        """
        batch = SettlementBatch(batch_id=batch_id) if batch_id else SettlementBatch()

        # Store source transactions
        for txn in transactions:
            batch.add_transaction(txn)

        # Calculate net positions
        positions = self.calculate_net_positions(transactions)
        batch.net_positions = positions

        # Generate minimal settlement instructions
        instructions = self.generate_settlements(positions)
        for instr in instructions:
            batch.add_instruction(instr)

        return batch

    def run_settlement_from_dicts(
        self, transactions: List[Dict], batch_id: str = None
    ) -> SettlementBatch:
        """
        Run settlement from transaction dictionaries.

        Args:
            transactions: List of dicts with 'payer', 'payee', 'amount' keys
            batch_id: Optional batch identifier

        Returns:
            SettlementBatch containing positions and instructions
        """
        txn_objects = [
            Transaction(
                payer=t["payer"],
                payee=t["payee"],
                amount=Decimal(str(t["amount"]))
            )
            for t in transactions
        ]
        return self.run_settlement(txn_objects, batch_id)


def demo():
    """Demonstrate the netting engine with sample data."""
    # Sample transactions from the problem statement
    sample_transactions = [
        {"payee": "BoA", "amount": 132, "payer": "Chase"},
        {"payee": "BoA", "amount": 827, "payer": "Chase"},
        {"payee": "Wells Fargo", "amount": 751, "payer": "BoA"},
        {"payee": "BoA", "amount": 585, "payer": "Chase"},
        {"payee": "Chase", "amount": 877, "payer": "Wells Fargo"},
        {"payee": "Wells Fargo", "amount": 157, "payer": "Chase"},
        {"payee": "Wells Fargo", "amount": 904, "payer": "Chase"},
        {"payee": "Chase", "amount": 548, "payer": "Wells Fargo"},
        {"payee": "Chase", "amount": 976, "payer": "BoA"},
    ]

    engine = NettingEngine()
    batch = engine.run_settlement_from_dicts(sample_transactions)

    print("Net Positions:")
    print("-" * 40)
    for bank, position in sorted(batch.net_positions.items()):
        if position > 0:
            print(f"  {bank}: +{position} (receives)")
        elif position < 0:
            print(f"  {bank}: {position} (pays)")
        else:
            print(f"  {bank}: 0 (flat)")

    print(f"\nSettlement Instructions ({len(batch.instructions)} transfers):")
    print("-" * 40)
    for instr in batch.instructions:
        print(f"  {instr}")

    print(f"\nNetting Statistics:")
    print("-" * 40)
    print(f"  Original transactions: {len(batch.source_transactions)}")
    print(f"  Gross volume: ${batch.total_gross_volume()}")
    print(f"  Net volume: ${batch.total_net_volume()}")
    print(f"  Netting efficiency: {batch.netting_efficiency():.1f}%")


if __name__ == "__main__":
    demo()

File: financial-clearing-house/src/java/com/clearinghouse/service/NettingEngine.java
package com.clearinghouse.service;

import com.clearinghouse.model.SettlementInstruction;
import com.clearinghouse.model.Transaction;

import java.math.BigDecimal;
import java.util.*;

/**
 * Multilateral netting engine that minimizes settlement transfers.
 *
 * Algorithm:
 * 1. Calculate net position per bank (sum of all inflows - outflows)
 * 2. Partition banks into creditors (net > 0) and debtors (net < 0)
 * 3. Greedily match largest creditor with largest debtor
 * 4. Generate settlement instruction for the matched amount
 * 5. Update positions and repeat until all positions are zero
 *
 * This achieves the minimum number of transfers (N-1 for N non-zero banks)
 * when there are no transfer fees to optimize for.
 */
public class NettingEngine {

    /**
     * Result of the netting process.
     */
    public static class NettingResult {
        private final Map<String, BigDecimal> netPositions;
        private final List<SettlementInstruction> instructions;
        private final BigDecimal grossVolume;
        private final BigDecimal netVolume;

        public NettingResult(Map<String, BigDecimal> netPositions,
                            List<SettlementInstruction> instructions,
                            BigDecimal grossVolume) {
            this.netPositions = netPositions;
            this.instructions = instructions;
            this.grossVolume = grossVolume;
            this.netVolume = instructions.stream()
                    .map(SettlementInstruction::getAmount)
                    .reduce(BigDecimal.ZERO, BigDecimal::add);
        }

        public Map<String, BigDecimal> getNetPositions() {
            return netPositions;
        }

        public List<SettlementInstruction> getInstructions() {
            return instructions;
        }

        public BigDecimal getGrossVolume() {
            return grossVolume;
        }

        public BigDecimal getNetVolume() {
            return netVolume;
        }

        public double getNettingEfficiency() {
            if (grossVolume.compareTo(BigDecimal.ZERO) == 0) {
                return 100.0;
            }
            return (1.0 - netVolume.doubleValue() / grossVolume.doubleValue()) * 100.0;
        }
    }

    /**
     * Calculate the net position for each bank.
     *
     * @param transactions List of transactions to process
     * @return Map from bank ID to net position (positive = receive, negative = pay)
     */
    public Map<String, BigDecimal> calculateNetPositions(List<Transaction> transactions) {
        Map<String, BigDecimal> positions = new HashMap<>();

        for (Transaction txn : transactions) {
            // Payer loses money
            positions.merge(txn.getPayer(), txn.getAmount().negate(), BigDecimal::add);
            // Payee gains money
            positions.merge(txn.getPayee(), txn.getAmount(), BigDecimal::add);
        }

        return positions;
    }

    /**
     * Validate that all positions sum to zero (system is balanced).
     *
     * @param positions Net positions per bank
     * @throws IllegalStateException if the system is unbalanced
     */
    public void validatePositions(Map<String, BigDecimal> positions) {
        BigDecimal total = positions.values().stream()
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        if (total.compareTo(BigDecimal.ZERO) != 0) {
            throw new IllegalStateException(
                    "System unbalanced! Positions sum to " + total + ", expected 0. " +
                    "This indicates lost or created money.");
        }
    }

    /**
     * Generate minimal settlement instructions using greedy matching.
     *
     * Uses a heap-based approach to always match the largest creditor
     * with the largest debtor for optimal efficiency.
     *
     * @param positions Net position per bank
     * @return List of SettlementInstruction objects (minimal set)
     */
    public List<SettlementInstruction> generateSettlements(Map<String, BigDecimal> positions) {
        validatePositions(positions);

        // Max-heap for creditors (positive positions)
        // Comparator: larger amount first
        PriorityQueue<Map.Entry<String, BigDecimal>> creditors = new PriorityQueue<>(
                (a, b) -> b.getValue().compareTo(a.getValue())
        );

        // Max-heap for debtors (negative positions, compare absolute values)
        // Comparator: larger debt (more negative) first
        PriorityQueue<Map.Entry<String, BigDecimal>> debtors = new PriorityQueue<>(
                (a, b) -> a.getValue().compareTo(b.getValue())
        );

        for (Map.Entry<String, BigDecimal> entry : positions.entrySet()) {
            if (entry.getValue().compareTo(BigDecimal.ZERO) > 0) {
                creditors.add(new AbstractMap.SimpleEntry<>(entry.getKey(), entry.getValue()));
            } else if (entry.getValue().compareTo(BigDecimal.ZERO) < 0) {
                debtors.add(new AbstractMap.SimpleEntry<>(entry.getKey(), entry.getValue()));
            }
        }

        List<SettlementInstruction> instructions = new ArrayList<>();

        while (!creditors.isEmpty() && !debtors.isEmpty()) {
            Map.Entry<String, BigDecimal> creditor = creditors.poll();
            Map.Entry<String, BigDecimal> debtor = debtors.poll();

            BigDecimal creditAmount = creditor.getValue();
            BigDecimal debtAmount = debtor.getValue().abs();

            // Settlement amount is the minimum of the two
            BigDecimal settlementAmount = creditAmount.min(debtAmount);

            // Generate settlement instruction
            SettlementInstruction instruction = new SettlementInstruction(
                    debtor.getKey(),   // payer (debtor)
                    creditor.getKey(), // payee (creditor)
                    settlementAmount
            );
            instructions.add(instruction);

            // Update remaining positions
            BigDecimal remainingCredit = creditAmount.subtract(settlementAmount);
            BigDecimal remainingDebt = debtAmount.subtract(settlementAmount);

            if (remainingCredit.compareTo(BigDecimal.ZERO) > 0) {
                creditors.add(new AbstractMap.SimpleEntry<>(creditor.getKey(), remainingCredit));
            }
            if (remainingDebt.compareTo(BigDecimal.ZERO) > 0) {
                debtors.add(new AbstractMap.SimpleEntry<>(debtor.getKey(), remainingDebt.negate()));
            }
        }

        return instructions;
    }

    /**
     * Run the full settlement process for a list of transactions.
     *
     * @param transactions List of transactions to settle
     * @return NettingResult containing positions and instructions
     */
    public NettingResult runSettlement(List<Transaction> transactions) {
        // Calculate gross volume
        BigDecimal grossVolume = transactions.stream()
                .map(Transaction::getAmount)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // Calculate net positions
        Map<String, BigDecimal> positions = calculateNetPositions(transactions);

        // Generate minimal settlement instructions
        List<SettlementInstruction> instructions = generateSettlements(positions);

        return new NettingResult(positions, instructions, grossVolume);
    }

    /**
     * Print the netting result to stdout.
     */
    public void printResult(NettingResult result, int originalTxnCount) {
        System.out.println("\nNet Positions:");
        System.out.println("-".repeat(40));

        List<String> sortedBanks = new ArrayList<>(result.getNetPositions().keySet());
        Collections.sort(sortedBanks);

        for (String bank : sortedBanks) {
            BigDecimal position = result.getNetPositions().get(bank);
            String status = position.compareTo(BigDecimal.ZERO) > 0 ? "receives" :
                           position.compareTo(BigDecimal.ZERO) < 0 ? "pays" : "flat";
            System.out.println("  " + bank + ": " + position + " (" + status + ")");
        }

        System.out.println("\nSettlement Instructions (" + result.getInstructions().size() + " transfers):");
        System.out.println("-".repeat(40));
        int i = 1;
        for (SettlementInstruction instr : result.getInstructions()) {
            System.out.println("  " + i++ + ". " + instr);
        }

        System.out.println("\nStatistics:");
        System.out.println("-".repeat(40));
        System.out.println("  Original transactions: " + originalTxnCount);
        System.out.println("  Gross volume: $" + result.getGrossVolume());
        System.out.println("  Net volume: $" + result.getNetVolume());
        System.out.printf("  Netting efficiency: %.1f%%\n", result.getNettingEfficiency());
    }
}

